core/servicedata/service.go

```diff
@@ -0,0 +1,111 @@
+package servicedata
+
+import (
+	"context"
+	"fmt"
+	"strings"
+
+	"github.com/goto/shield/core/project"
+	"github.com/goto/shield/core/relation"
+	"github.com/goto/shield/core/resource"
+	"github.com/goto/shield/core/user"
+	"github.com/goto/shield/internal/schema"
+	"github.com/goto/shield/pkg/uuid"
+)
+
+const keyNamespace = "shield/servicedata_key"
+
+type ResourceService interface {
+	Create(ctx context.Context, res resource.Resource) (resource.Resource, error)
+}
+
+type RelationService interface {
+	Create(ctx context.Context, rel relation.RelationV2) (relation.RelationV2, error)
+}
+
+type ProjectService interface {
+	Get(ctx context.Context, idOrSlug string) (project.Project, error)
+}
+
+type UserService interface {
+	FetchCurrentUser(ctx context.Context) (user.User, error)
+}
+
+type Service struct {
+	repository      Repository
+	resourceService ResourceService
+	relationService RelationService
+	projectService  ProjectService
+	userService     UserService
+}
+
+func NewService(repository Repository, resourceService ResourceService, relationService RelationService, projectService ProjectService, userService UserService) *Service {
+	return &Service{
+		repository:      repository,
+		resourceService: resourceService,
+		relationService: relationService,
+		projectService:  projectService,
+		userService:     userService,
+	}
+}
+
+func (s Service) CreateKey(ctx context.Context, key Key) (Key, error) {
+	// check if key contains ':'
+	if strings.Contains(key.Key, ":") || key.Key == "" {
```

Comment:
We can add this constrain later after we migrate the existing key I believe

```suggestion
	if key.Key == "" {
```